src/zenml/utils/notebook_utils.py

```diff
@@ -0,0 +1,249 @@
+#  Copyright (c) ZenML GmbH 2024. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Notebook utilities."""
+
+import json
+import os
+from typing import Any, Callable, Dict, Optional, TypeVar, Union
+
+from IPython import get_ipython
+
+from zenml.constants import ENV_ZENML_NOTEBOOK_PATH
+from zenml.environment import Environment
+from zenml.logger import get_logger
+
+ZENML_NOTEBOOK_CELL_ID_ATTRIBUTE_NAME = "__zenml_notebook_cell_id__"
+
+DEFAULT_NOTEBOOK_NAME = "test.ipynb"
+
+logger = get_logger(__name__)
+
+_UNINITIALIZED = "uninitialized"
+_ACTIVE_NOTEBOOK_PATH: Optional[str] = _UNINITIALIZED
+
+O = TypeVar("O", bound=Any)
+
+
+def get_active_notebook_path() -> Optional[str]:
+    """Get path of the active notebook.
+
+    Returns:
+        Path of the active notebook.
+    """
+    if not Environment.in_notebook():
+        return None
+
+    global _ACTIVE_NOTEBOOK_PATH
+
+    if _ACTIVE_NOTEBOOK_PATH == _UNINITIALIZED:
+        import ipynbname
+
+        from zenml.utils import source_utils
+
+        if path := os.environ.get(ENV_ZENML_NOTEBOOK_PATH, None):
+            logger.info(
+                "Using notebook path from environment variable: %s", path
+            )
+
+            _ACTIVE_NOTEBOOK_PATH = path
+        else:
+            from IPython import get_ipython
+
+            ip = get_ipython()
+            if "__vsc_ipynb_file__" in ip.user_ns:
+                _ACTIVE_NOTEBOOK_PATH = ip.user_ns["__vsc_ipynb_file__"]
+            else:
+                try:
+                    _ACTIVE_NOTEBOOK_PATH = str(ipynbname.path())
+                except FileNotFoundError:
+                    logger.warning(
+                        "Unable to detect active notebook. You can use the %s "
+                        "environment variable to manually specify a path to the "
+                        "notebook that you're currently using.",
+                        ENV_ZENML_NOTEBOOK_PATH,
```

Comment:
Minor: we mostly use `f-string` in the code.

Reply:
I did some extra reading on this and have to admit that this is desirable approach 🙂 
In this case it would make sense to refactor a bit, since we use f-strings heavily in logging.